In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [29]:
data=pd.read_csv('mercedesbenz.csv',usecols=['X1','X2','X3','X4','X5','X6'])
data.head()

X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d

In [30]:
data.shape

(4209, 6)

In [31]:
#fetching counts of all the uniques labels for each columns
for x in data.columns:
    print(x, "-",len(data[x].unique())," labels")

X1 - 27  labels
X2 - 44  labels
X3 - 7  labels
X4 - 4  labels
X5 - 29  labels
X6 - 12  labels


In [32]:
newData=pd.get_dummies(data,drop_first=True)

In [33]:
newData.shape

(4209, 117)

##### we can observe that from just 6 categorical variable we have converted it to 117 columns increasing the complexity of feature Instead we can do is make one binary variable for each of 10 most frequent labels only for each column, that is equivalent to grouping all the other variables in new category which will be dropped in our case. If 10 new most frequent is present than (1) else (0)

In [47]:

# lets find out top 10 labels for X2 variable
data.X2.value_counts().sort_values(ascending=False).head(10)

as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
Name: X2, dtype: int64

In [35]:
#fetching & storing top 10 most frequent labels of variable 'X2'
top10=[x for x in data.X2.value_counts().sort_values(ascending=False).head(10).index]
top10

['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e']

In [40]:
# Using np.where to encode with 1 if current label(iterator value) matches with label in that particular variable.
for labels in top10:
    data[labels]=np.where(data['X2']==labels,1,0)

In [41]:
data[['X2']+top10].head(20)

X2  as  ae  ai  m  ak  r  n  s  f  e
0   at   0   0   0  0   0  0  0  0  0  0
1   av   0   0   0  0   0  0  0  0  0  0
2    n   0   0   0  0   0  0  1  0  0  0
3    n   0   0   0  0   0  0  1  0  0  0
4    n   0   0   0  0   0  0  1  0  0  0
5    e   0   0   0  0   0  0  0  0  0  1
6    e   0   0   0  0   0  0  0  0  0  1
7   as   1   0   0  0   0  0  0  0  0  0
8   as   1   0   0  0   0  0  0  0  0  0
9   aq   0   0   0  0   0  0  0  0  0  0
10   r   0   0   0  0   0  1  0  0  0  0
11   e   0   0   0  0   0  0  0  0  0  1
12  ai   0   0   1  0   0  0  0  0  0  0
13   e   0   0   0  0   0  0  0  0  0  1
14  as   1   0   0  0   0  0  0  0  0  0
15  ak   0   0   0  0   1  0  0  0  0  0
16   e   0   0   0  0   0  0  0  0  0  1
17  as   1   0   0  0   0  0  0  0  0  0
18   m   0   0   0  1   0  0  0  0  0  0
19  ak   0   0   0  0   1  0  0  0  0  0

In [64]:
# lets make the complete function for one hot encoding

def one_hot_encode(dataframe, var, top_x_value):
    
    #fetching & storing top 10 most frequent labels of given variable 
    topLabels=[x for x in dataframe[var].value_counts().sort_values(ascending=False).head(top_x_value).index]
    
    #Using np.where to encode with 1 if current label(iterator value) matches with label in that particular variable.
    for labels in topLabels:
        dataframe[var+"_"+labels]=np.where(dataframe[var]==labels,1,0)
        

mercedes_data=pd.read_csv('mercedesbenz.csv',usecols=['X1','X2','X3','X4','X5','X6'])

#mercedes_data['X2'].value_counts().sort_values(ascending=False).head(10)

one_hot_encode(mercedes_data, 'X2', 10)

In [65]:
mercedes_data

X1  X2 X3 X4  X5 X6  X2_as  X2_ae  X2_ai  X2_m  X2_ak  X2_r  X2_n  X2_s  \
0     v  at  a  d   u  j      0      0      0     0      0     0     0     0   
1     t  av  e  d   y  l      0      0      0     0      0     0     0     0   
2     w   n  c  d   x  j      0      0      0     0      0     0     1     0   
3     t   n  f  d   x  l      0      0      0     0      0     0     1     0   
4     v   n  f  d   h  d      0      0      0     0      0     0     1     0   
...  ..  .. .. ..  .. ..    ...    ...    ...   ...    ...   ...   ...   ...   
4204  s  as  c  d  aa  d      1      0      0     0      0     0     0     0   
4205  o   t  d  d  aa  h      0      0      0     0      0     0     0     0   
4206  v   r  a  d  aa  g      0      0      0     0      0     1     0     0   
4207  r   e  f  d  aa  l      0      0      0     0      0     0     0     0   
4208  r  ae  c  d  aa  g      0      1      0     0      0     0     0     0   

      X2_f  X2_e  
0        0     0  
1        0     0  
2        0     0  
3        0     0  
4        0     0  
...    ...   ...  
4204     0     0  
4205     0     0  
4206     0     0  
4207     0     1  
4208     0     0  

[4209 rows x 16 columns]